kusto/ingest/result.go

```diff
@@ -0,0 +1,149 @@
+package ingest
+
+import (
+	"context"
+	"time"
+
+	"github.com/Azure/azure-kusto-go/kusto/ingest/internal/properties"
+	"github.com/Azure/azure-kusto-go/kusto/ingest/internal/resources"
+	"github.com/Azure/azure-kusto-go/kusto/ingest/internal/status"
+)
+
+// Result provides a way for users track the state of ingestion jobs.
+type Result struct {
+	record        StatusRecord
+	tableClient   *status.TableClient
+	reportToTable bool
+	reportToQueue bool
+}
+
+// newResult creates an initial ingestion status record.
+func newResult() *Result {
+	ret := &Result{}
+
+	ret.record = newStatusRecord()
+	return ret
+}
+
+// putErr sets the record to a failure state and adds the error to the record details.
+func (r *Result) putErr(err error) *Result {
+	return r.putErrStr(err.Error())
+}
+
+// putErrStr sets the record to a failure state and adds the error to the record details.
+func (r *Result) putErrStr(err string) *Result {
+	r.record.Status = ClientError
+	r.record.FailureStatus = Permanent
+	r.record.Details = err
+
+	return r
+}
+
+// putProps sets the record to a failure state and adds the error to the record details.
+func (r *Result) putProps(props properties.All) *Result {
+	r.reportToTable = props.Ingestion.ReportMethod == properties.ReportStatusToTable || props.Ingestion.ReportMethod == properties.ReportStatusToQueueAndTable
+	r.record.FromProps(props)
+
+	return r
+}
+
+// putQueued sets the initial success status depending on status reporting state
+func (r *Result) putQueued(mgr *resources.Manager) *Result {
+	// If not checking status, just return queued
+	if !r.reportToTable {
+		r.record.Status = Queued
+		return r
+	}
+
+	// Get table URI
+	resources, err := mgr.Resources()
+	if err != nil {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Transient
+		r.record.Details = "Failed getting status table URI: " + err.Error()
+		return r
+	}
+
+	if len(resources.Tables) == 0 {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Transient
+		r.record.Details = "Ingestion resources do not include a status table URI: " + err.Error()
+		return r
+	}
+
+	// create a table client
+	client, err := status.NewTableClient(*resources.Tables[0])
+	if err != nil {
+		r.record.Status = StatusRetrievalFailed
+		r.record.FailureStatus = Transient
+		r.record.Details = "Failed Creating a Status Table client: " + err.Error()
+		return r
+	}
+
+	// Write initial record
+	r.record.Status = Pending
+	recordMap := r.record.ToMap()
+	err = client.WriteIngestionStatus(r.record.IngestionSourceID.String(), recordMap)
+	if err != nil {
+		r.putErr(err)
+	} else {
+		r.tableClient = client
+	}
+
+	return r
+}
+
+// Wait returns a channel that can be checked for ingestion results.
+// In order to check actual status please use the IngestionStatus option when ingesting data.
+func (r *Result) Wait(ctx context.Context) chan StatusRecord {

```

Comment:
So I want to go back to this.  I think StatusRecord should probably be internal/  and this should probably return error.

That error can be a StatusRecord, if status record implements the "Error() string" method.  Then we can return a mix of errors.  There is no use for StatusRecord itself.  

There are only a few states that the user will receive, which basically say: pass/fail/partial.  That is all the user is going to care about + the text message.

And if we changed putErr() to store the actual error when doing client errors, you could return a client specific error so they could check if they should retry easily.  Or we could return (*Result, error) where error is always client errors (honestly I like this better).  

```go
type ErrClient struct {
  Msg string
}
func (e ErrClient) Error() string {
  return e.Msg
}
```

But I want to get rid of access to StatusRecord and probably do a statusRecord which implements Error(). 

Also, going to put nots on StatusRecord below.

Reply:
No, because then the person could tease the StatusRecord back by doing:

err := blah.Wait()
if sr, ok := err.(StatusRecord); ok {
  // Do stuff with StatusRecord
}

So StatusRecord would also need to become statusRecord + ToError() -> Error(), preventing them from doing that.  Or it would need to be in an internal/ directory.

You can leak a private type with public methods which gives them access to public methods but not the type.

But I'm not in love with this.  

Here's what I think we should do:

```go
FromReader(ctx context.Context, reader io.Reader, options ...FileOption) *Result
```
should become
```go
FromReader(ctx context.Context, reader io.Reader, options ...FileOption) (*Result, error)
```

The error would be a clientError from status.go

This differentiates clearly our errors that aren't directly ingestion data errors.

Status records should move to internal/ or become statusRecord

You can keep ToError() and just return that in *Result.Wait().

If later we want to distinguish between these error types (and I can't see why we would programmically care), we can still do that.

Reply:
Thinking of our goals from today, I think what I'm proposing below will meet these:
- Hide the implementation (StatusRecord) from the user
- Allow the user to know if they can retry an error
- The error must have diagnostic information from StatusRecord for support calls
- Allow the possibility, in the future, of exposing StatusRecord via Wait(), if required

## Proposed changes

### StatusRecord to become private
StatusRecord -> statusRecord

### Make statusRecord implement the built in error type

statusRecord.Error() ->
```go
// Error implements error.Error().
func (r *statusRecord) Error() string {
	return r.String()
}
```
Notes:
	- We will only return statusRecord as an error if the status was not Succeeded.  Any other status is a failure and they will never see status like Queued.
	- When there is an error, it will give them the string output of the StatusRecord for support calls.
	- This always gives us the option, at a later date, to make StatusRecord public and expose it to the user

### Change Result.Wait to return a chan error
Result.Wait() ->
```go
func (r *Result) Wait(ctx context.Context) chan error {
...
}
```
Notes:
	- If there is an error, you return the statusRecord, which now implements error
	- Otherwise you just return nil


### Provide the ability to know if an error can be retried
```go
func CanRetry(e error) bool{
	if sr, ok := e.(statusRecord); ok {
	    switch sr.FailureStatus{
	    case Unknown, Permanent:
		    return false
	    }
	    return true
        }
        if errors.Is(e, context.Canceled || errors.Is(e, context.DeadlineExceeded) {
            return true
        }
        return false
}
```

## Final notes
Notes: 
	- FailureStatusCode would become failureStatusCode and all instances would be private (we can always change that later)
	- StatusCode would become private and all instances (we can always change that later)
	- ToMap() should become toMap() (I didn't see any reason for it to be public, but I could be wrong)
	- FromMap() should become fromMap() (same)